Add support for SPI emulators

sjg20 opened the pull request:
Add an emulation controller which routes SPI traffic to an attached emulator.

This allows drivers for SPI peripherals to be tested on systems that don't have that peripheral attached, with the emulator handling the SPI traffic.

As an example, add an emulator for the Boshe BMI160 accelerometer and a test for that,

sjg20 committed:
tests: drivers: sensor: Move current test into generic/

The existing tests is for generic sensor drivers. To make room for a
driver-specific one, move it into a subdirectory.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
tests: drivers: sensor: Add an BMI160 accelerometer test

Using the SPI emulator we can create a simple test for the BMI160
driver. Add this an enable it for native_posix.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
drivers: sensor: bmi160: Use more constants for reg access

At present register access is a mix of constants and open-coded values
in the driver. Add a few more constants to clean this up.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
drivers: sensor: bmi160: Add a constant for number of axes

Define a constant instead of using an open-coded value.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
drivers: sensor: bmi160: Move the ready bit into the header

This should be accessible to all files that need register access. Move
it into the header.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
emul: spi: Add support for SPI emulators

Add an emulation controller which routes SPI traffic to an attached
emulator. Only one emulator is supported per bus at present, since
chip-selction functionality is not present.

This allows drivers for SPI peripherals to be tested on systems that
don't have that peripheral attached, with the emulator handling the SPI
traffic.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
emul: Add an emulator for the Bosch BMI160 accelerometer

This emulator supports enable functionality to start up the device and
read a few samples. It connects itself to any BMI160 device it finds in
the device tree. The SPI emulation controller driver is used to direct
SPI messages from the BMI160 driver to the BMI160 emulator.

Add a few more definitions to the header file, as needed.

Signed-off-by: Simon Glass <sjg@chromium.org>

jettr commented on boards/posix/native_posix/native_posix.dts:
```diff
@@ -91,6 +92,16 @@
 		label = "I2C_0";
 	};
 
+	spi0: spi@200 {
```
Does the emulator need a base memory address (of 200)? If not, can we drop it?

sjg20 replied:
No it doesn't need it, but peripherals are supposed to be at a particular address - see I2C. That said I don't think it is essential so I'll drop it.

sjg20 replied:
Hmm actually I can't, since the SPI binding requires it.

jettr replied:
No worries, then. We can leave it

tbursztyka reviewed (dismissed):
quickly went through it: I missed where the actual drivers/sensors/bmi160 transaction end up to the emulator? could you pin point the file/line?

sjg20 commented on kernel/device.c:
```diff
@@ -31,6 +32,8 @@ extern uint32_t __device_busy_end[];
 #define DEVICE_BUSY_SIZE (__device_busy_end - __device_busy_start)
 #endif
 
+LOG_MODULE_REGISTER(DEVICE, 7);
```
I'm going to put this in a separate PR

sjg20 commented on drivers/spi/spi_emul.c:
```diff
@@ -0,0 +1,147 @@
+
+	emul = spi_emul_find(dev);
+	if (!emul) {
+		return -EIO;
```
Maybe, but it seems a bit odd to return that there is no device:

- clearly @dev is a device
- the emulator is (for now at least) not a device

This is akin to the hardware being disconnected, so it simply doesn't respond. Perhaps a better emulation would be to return 0 and do nothing? But I worry that it would be hard for people to debug.

What do you think?

jettr replied:
Ah, I see that more clearly now. I think `-EIO` makes sense

sjg20 commented on drivers/spi/spi_emul.c:
```diff
@@ -0,0 +1,147 @@
+		return ret;
+	}
+
+	return 0;
```
I guess this is OK since I haven't added any logging and it is pretty obvious that it can fail.

sjg20 commented on include/drivers/spi_emul.h:
```diff
@@ -0,0 +1,94 @@
+ * Copyright 2020 Google LLC
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
```
It is copied from from the I2C file since Nordic wrote some of the code. I suppose I can drop it since this is a new file

pabigot replied:
We aren't consistent in the placement.  Linux puts SPDX at the first line; that's where I put it.  Doxygen doesn't care (the files I remember seeing that start with the module purpose had doxygen comments `@file`).

It's relatively unusual to have a text that isn't related to copyright and SPDX in the same block.  If we were to standardize, I would have it first, with no other text.

Opinons aside, the count of number of occurrences of `SPDX` in a header file at each line strongly suggests copyright/spdx first is the de facto standard.
```
tirzah[133]$ git grep -n SPDX -- '*/*.h' | cut -d: -f2 | sort | uniq -c | sort -n -k2
       6 1
     11 2
     45 3
    967 4
    168 5
    113 6
     10 7
     65 8
     21 9
     49 10
     10 11
      1 12
      1 29
      1 36
```

jettr replied:
thanks for the moving the file brief and removing Nordic copyright

sjg20 commented on boards/posix/native_posix/native_posix.dts:
```diff
@@ -91,6 +92,16 @@
+		reg = <0x200 4>;
+		label = "SPI_0";
+	};
+
```
Please see https://github.com/zephyrproject-rtos/zephyr/issues/27531 for why
This is not just for running a single test. The idea is that you can emulate a whole system with this scheme. The I2C emulator driver is enabled for native_posix, so I think the SPI one should be too.

sjg20 commented on kernel/device.c:
```diff
@@ -84,17 +87,22 @@ const struct device *z_impl_device_get_binding(const char *name)
 			return dev;
 		}
 	}
+	LOG_DBG("   - not found");
```
I've moved this to another PR. The idea here is to catch problems no matter what code calls it

sjg20 committed:
emul: Add messages in emul_init_for_bus_from_list

Add a log message to indicate that a new emulator is attached. Also add
a message in the assert to make it clearer what has gone wrong.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 commented on subsys/emul/emul.c:
```diff
@@ -46,10 +46,12 @@ int emul_init_for_bus_from_list(const struct device *dev,
 	const struct emul_link_for_bus *const end =
 		cfg->children + cfg->num_children;
 
+	LOG_INF("Registering %d emulator(s) for %s", cfg->num_children,
```
done

sjg20 commented on subsys/emul/spi/emul_bmi160.c:
```diff
@@ -0,0 +1,265 @@
+	/** SPI emulator detail */
+	struct spi_emul emul;
+	/** BMI160 device being emulated */
+	const struct device *spi;
```
done

pabigot commented on subsys/emul/spi/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
```
I think it makes more sense to do:

    zephyr_include_directories_ifdef(CONFIG_EMUL_BMI160 ${ZEPHYR_BASE}/drivers/sensor)

than to move the private driver header into the global include directory.

sjg20 replied:
Does that mean:

zephyr_include_directories_ifdef(CONFIG_EMUL_BMI160 ${ZEPHYR_BASE}/drivers/sensor/bmi160)

and then using:

#include <bmi160.h>

in the emulator? If so, it works, but it seems confusing to include bmi160.h as a header file that I would expect to find in include/ but which is actually in drivers/sensor/bmi160/

pabigot replied:
Yes, that's what I meant.  I don't think you should expect to find headers that are intended for driver implementation sources (or emulator implementation sources) in the system include area.  The main include area should be for things that applications are expected to include, and not be cluttered with things they are not expected to include.

sjg20 replied:
OK done

pabigot reviewed (changes requested):
I'd like to see the unit address used to support multiple devices on the bus, just as it is for I2C.  That there's no simulated GPIO support right now shouldn't prevent that preparation, and I'm hoping #26484 makes progress early in 2.5 development (@cfriedt).

Also I'd like to avoid moving device-specific headers into the standard include hierarchy when they're not providing device-specific API for users of the devices.  Developers of emulators should be allowed to reference the driver implementation area to get what they need from existing headers.

The BMI160 emulator has more magic numbers than I'm happy with, but I won't block on that, except for the one I specifically called out.

sjg20 commented on subsys/emul/spi/emul_bmi160.c:
```diff
@@ -0,0 +1,265 @@
+	case BMI160_REG_GYR_CONF:
+		LOG_INF("   * gyr conf");
+		break;
+	case 0x7f:
```
done

sjg20 commented on drivers/spi/spi_emul.c:
```diff
@@ -0,0 +1,137 @@
+ * Find an emulator for a SPI bus
+ *
+ * At present only a single emulator is supported on the bus, since we do not
+ * support chip selects.
```
Hmmm. I took that out in the most recent version. I'll put it back in

sjg20 replied:
done

sjg20 commented on drivers/sensor/bmi160/bmi160.c:
```diff
@@ -12,13 +12,12 @@
 
 #include <init.h>
 #include <drivers/sensor.h>
+#include <drivers/sensor/bmi160.h>
```
Does linux have an emulator?

sjg20 commented on subsys/emul/spi/emul_bmi160.c:
```diff
@@ -0,0 +1,265 @@
+					reg_write(cfg, regn, val);
+				}
+				break;
+			case 12:
```
done

sjg20 commented:
Thanks for the review. This is updated now. I'll try fixing the tests.

pabigot reviewed (changes requested):
Please integrate the change that supports chip select.  Then I think this is OK.

sjg20 commented:
> Please integrate the change that supports chip select. Then I think this is OK.

I added chipsel - what else are you expecting? I can't actually implement it until the GPIO emulator is ready, as I understand it.

pabigot commented:
> I added chipsel - what else are you expecting? I can't actually implement it until the GPIO emulator is ready, as I understand it.

Many SPI peripherals don't use an external GPIO for chip select.  It's managed internal to the peripheral, which assigns the gpio based on pinmux.  The commit I referenced shows how to use the reg attribute from the devicetree to differentiate multiple emulator devices on the same bus.

pabigot commented:
> The commit I referenced shows how to use the reg attribute from the devicetree to differentiate multiple emulator devices on the same bus.

Apparently my comments didn't get recorded?  I was referencing https://github.com/pabigot/zephyr/commit/08801c1cd889cc2b9f68747eba5a74f28a5890f5

sjg20 commented:
> > I added chipsel - what else are you expecting? I can't actually implement it until the GPIO emulator is ready, as I understand it.
> 
> Many SPI peripherals don't use an external GPIO for chip select. It's managed internal to the peripheral, which assigns the gpio based on pinmux. The commit I referenced shows how to use the reg attribute from the devicetree to differentiate multiple emulator devices on the same bus.

OK I see. I actually completely missed the slave parameter in spi_config. I was looking at cs. I suppose slave is a reasonable name but chip-select seems better to me.

sjg20 commented:
> > The commit I referenced shows how to use the reg attribute from the devicetree to differentiate multiple emulator devices on the same bus.
> 
> Apparently my comments didn't get recorded? I was referencing [pabigot@08801c1](https://github.com/pabigot/zephyr/commit/08801c1cd889cc2b9f68747eba5a74f28a5890f5)

OK, got it. That we we hopefully don't have to do anything when the GPIO thing is ready

sjg20 commented:
> > > The commit I referenced shows how to use the reg attribute from the devicetree to differentiate multiple emulator devices on the same bus.
> > 
> > 
> > Apparently my comments didn't get recorded? I was referencing [pabigot@08801c1](https://github.com/pabigot/zephyr/commit/08801c1cd889cc2b9f68747eba5a74f28a5890f5)
> 
> OK, got it. That we we hopefully don't have to do anything when the GPIO thing is ready

done
